### DISCUSSION TITLE
Why not run kernel from flash on Pi Pico

### DISCUSSION THREAD

**Discussion by veremenko-y** [EtchedPixels/FUZIX#1053] [Q&A] [answered]
It's generally a question to @davidgiven, but I wanted to keep it public for others to chime in if needed.

I was wondering if you had specific reasons not to run the kernel from flash directly? It does free up quite a bit of RAM, and seems to work just fine for me, besides few UART glitches on startup.

**Comment by davidgiven** [comment, marked as answer]
Should one of those 'flash' be a 'RAM'?

I have vague memories that running from flash caused big problems with the secondary core and USB. Writing to the flash requires turning XIP off, which means going through a trampoline in RAM and stalling the secondary core until the write is complete. That would cause the USB thread to stop servicing requests and stuff would time out. I _think_ I tried putting the critical parts in RAM, but I clearly didn't make it work.

Ideally the entire kernel should be running from flash in order to maximise precious user-mode RAM...

**Reply by veremenko-y** [reply]
Yes, sorry. Second one should have been RAM. I've been stuck at home with flu, so my brain doesn't necessarily work well.

Now that I added an option to disable NAND storage, it should be possible to run form RAM without many issues. I need to track down the garbage characters in UART though.

    Devboot                                                                                
    264KiB total RAM, 233KiB available to processes (15 processes max)                     
    Enabling interrupts ... ok.
